api/session/session.go

```diff
@@ -1,50 +1,93 @@
 type Service struct {
 	timeout time.Duration
 	store   api.StorageService
+	log     api.LogService
 }
 
 // NewSessionService returns a SessionService
-func NewSessionService(timeout time.Duration, store api.StorageService) *Service {
+func NewSessionService(timeout time.Duration, store api.StorageService, log api.LogService) *Service {
 	return &Service{
 		timeout,
 		store,
+		log,
 	}
 }
 
-// UserDidAuthenticate returns a session key and an error if applicable
-func (s Service) UserDidAuthenticate(accountID int) (string, error) {
-	sessionKey := uuid.New().String()
+// generateSessionKey generates a cryptographically random session key
+func generateSessionKey() (string, error) {
+	secureBytes := securecookie.GenerateRandomKey(32)
+	if secureBytes == nil {
+		return "", errors.New("Failed to generate random data for a key")
+	}
 
-	// TODO: add tests to sanity check time edge cases / that time.Time is the right type to use here
-	expirationDate := time.Now().Add(s.timeout)
+	secureString := hex.EncodeToString(secureBytes)
 
-	createErr := s.store.CreateOrUpdateSession(accountID, sessionKey, expirationDate)
+	return secureString, nil
+
+}
 
-	// TODO: update accounts table with login datetime
+func hashSessionKey(sessionKey string) string {
+	hashed := sha512.Sum512([]byte(sessionKey))
+	hexEncoded := hex.EncodeToString(hashed[:])
+	return hexEncoded[:12]
+}
+
+// UserDidAuthenticate returns a session key and an error if applicable
+func (s Service) UserDidAuthenticate(accountID int, sessionIndex sql.NullString) (string, error) {
+	sessionKey, keyErr := generateSessionKey()
+	if keyErr != nil {
+		return "", keyErr
+	}
 
+	createErr := s.store.CreateOrUpdateSession(accountID, sessionKey, sessionIndex, s.timeout)
 	if createErr != nil {
 		return "", createErr
 	}
+	hashedSessionKey := hashSessionKey(sessionKey)
+	s.log.AddField("session_hash", hashedSessionKey)
+	s.log.Info(api.SessionCreated, api.LogFields{})
 
 	return sessionKey, createErr
 }
 
 // GetAccountIfSessionIsValid returns an Account if the session key is valid and an error otherwise
-func (s Service) GetAccountIfSessionIsValid(sessionKey string) (api.Account, error) {
-	return s.store.FetchSessionAccount(sessionKey)
+func (s Service) GetAccountIfSessionIsValid(sessionKey string) (api.Account, api.Session, error) {
+	account, session, fetchErr := s.store.ExtendAndFetchSessionAccount(sessionKey, s.timeout)
+	if fetchErr != nil {
+		if fetchErr == api.ErrSessionExpired {
+			s.log.Info(api.SessionExpired, api.LogFields{})
+		} else if fetchErr == api.ErrValidSessionNotFound {
+			s.log.Info(api.SessionDoesNotExist, api.LogFields{})
+		}
+
+		return api.Account{}, api.Session{}, fetchErr
+	}
+
+	return account, session, nil
 }
 
 // UserDidLogout attempts to end the session and returns an error on failure
 func (s Service) UserDidLogout(sessionKey string) error {
-	return s.store.DeleteSession(sessionKey)
+	delErr := s.store.DeleteSession(sessionKey)
```

Comment:
Do you have a strategy in mind for where to call `DeleteSession` for general clean up? i.e., avoid perpetual expansion of the session table with stale rows?

Reply:
I'll make sure that we are deleting rows from this table when DELETE account is called.

Reply:
But the real answer is no, I didn't really have a plan for this. The table is limited to the total number of users in the system so once the system has enough users that this is a problem it should be trivial to delete the old ones.